### ISSUE TITLE
params on img tag

### ISSUE THREAD

**Issue by MissAllSunday** [SimpleMachines/SMF#3434]
On Max display width of posted or attached images and Max display height of posted or attached images put something different than 0

now use the img bbc tag on an image that surpasses those sizes

BBC is correctly modified:

`[img width=373 height=400]http://i.imgur.com/7NjzUo8.jpg[/img]`

But it gets parse as:

`<img src="http://width=373 height=400]http://i.imgur.com/7NjzUo8.jpg" alt="" title="" &nbsp;="" class="bbc_img resized" style="cursor: pointer;">`

Trying to quote an already quoted text result in an attach BBC tag around the inner quote link:

```
[attach]Quote from: Simple Machines on Mar 22, 2016, 04:13 pm[/attach]
Quote
Welcome to Simple Machines Forum!

```

**Comment by MissAllSunday** [comment]
It seems this commit causes these issues: https://github.com/SimpleMachines/SMF2.1/commit/73aa070e49d217fb8c050a508eda8f041e7006d1

@Sesquipedalian do you have an idea why this is happening?

**Comment by Sesquipedalian** [comment]
Not off the top of my head, but I can look into it. It'll be a few days before I can get to it, though.

**Comment by Sesquipedalian** [comment]
Using an out of date branch that includes 73aa070, I can reproduce the bug with the img bbc, but not the quote --> attach bug. I'll look into what is going on with img parameters, but I think we'll need to look elsewhere to find the cause of the quote bug.

**Comment by Sesquipedalian** [comment]
I'm unable to reproduce the quote → attach bug using any version of the release-2.1 branch. Not sure why you are seeing that, @MissAllSunday, or what to do about it.

**Comment by MissAllSunday** [comment]
There are still issues with the way you are extracting the params on an unparsed_equals_content tag.

 I'm testing the attach BBC tag and it doesn't get parsed properly, if I only add a single atach bbc it doesn't get parsed at all  and if I add two or more then the parsed data contains HTML and the entire tag before the ID:

var_dump($attachID);  inside validate closure in attach tag:

`<br><br>[attach name=image.jpg type=image/jpeg]23`

**Comment by Sesquipedalian** [comment]
I cannot reproduce that after applying #3447, @MissAllSunday. Here's the text of a post I tried with:

```
This post has some image attachments.

Here it is:
[attach name=darkness.jpg type=image/jpeg]3[/attach]

Here is another:
[attach name=dendra.jpg type=image/jpeg]9[/attach]

Here is a third:
[attach name=jon+icon.jpg type=image/jpeg]13[/attach]
```

I added the following inside the validate section of the attach BBC:

```
echo '<pre>'; var_dump($attachID); echo '</pre>';
```

All three images rendered correctly, and the var_dump spits out the $attachID values as just integers. You can see the results at http://stovell.noip.me/~jon/index.php?topic=2.0

If you are still seeing problems, can you give me an example post that creates the issue?

**Comment by MissAllSunday** [comment]
It happens with whatever message that contains an attach tag.  I can reproduce it on a fresh copy with php 5.6 and on my site: http://missallsunday.com/blog/something_think_752.0.html with 5.4 and 7

**Comment by Sesquipedalian** [comment]
Pop open Subs.php and do a search for `$blob_counter`. It should be set to an initial value of 1 in the code. The behaviour on your page looks _exactly_ like what happens if the value is set to 0. Changing that 0 to a 1 is what #3447 does, so if your install has a 0 then you must be using some branch that doesn't include #3447. If the initial value really is 1 and you are nevertheless seeing that behaviour, then something weird is going on and I will need to work with you more to figure it out.

**Comment by MissAllSunday** [comment]
Yes.... if I'm reporting this is because I'm already using the fix you provided.  https://github.com/MissAllSunday/SMF2.1/blame/stuff/Sources/Subs.php#L2051

I just tested this on a fresh set of files on local and my server with the same result. The attach tag is completely bypassed if theres only one tag and if theres two or more the parsed reult contains the whole attach tag plus its ID.

This issue started to appear after the way the params gets parsed got changed.

**Comment by Sesquipedalian** [comment]
Hmm. Thanks, @MissAllSunday. It is quite strange that this is happening with your install but not mine. Once I am back at my computer I will dig in some more to try to reproduce the behaviour you are seeing and find the cause for the difference. 

In the meantime, can you tell me what behaviour you see if you roll back to 97876e7 and then manually change that 0 to a 1? It might or might not make any difference to what you are observing, but either way that info should help me narrow down the list of possible causes for the discrepancy between the behaviour you are getting and what I am getting.

**Comment by Sesquipedalian** [comment]
Okay, I just did another completely fresh install, and now I am seeing the same behaviour as you, @MissAllSunday. I probably forgot to switch to the right branch or something last time. :)

**Comment by Sesquipedalian** [comment]
Okay, so the problem with the `attach` tag appears to be that f81e4bb changed the tag type to `unparsed_equals_content`, which expects the form `[tag=...]content[/tag]`, but the code being inserted into the post still takes the form of an `unparsed_content` tag type (i.e. `[attach name=attachment.jpg type=image/jpeg]3[/attach]`). From what I can tell, f81e4bb was intended to fix #3381, but the real cause of #3381 was fixed in a1d1e88 (#3447). Reverting f81e4bb (and 91587e6, which modifies and depends on f81e4bb) seems to have fixed it all.
PR: #3457
Demo: http://stovell.noip.me/~jon/foo/index.php?topic=2.0

**Comment by MissAllSunday** [comment]
https://github.com/SimpleMachines/SMF2.1/commit/f81e4bbe23e104c6e876ef950af4dcf700b9a094   was made way before I notice the attach and img issues, it was done due to changes on the JS for correctly handling attach tag on the editor's WYSIWYG mode.

If we revert that back then the attach on that mod will be broken again.

**Comment by Sesquipedalian** [comment]
I'm just going by the bug report in #3381. @PortaMx's initial description of the issue with the attach tag in that report matches the set of symptoms described in this thread regarding the same tag. If you say f81e4bb was meant to address something more than that or other than that, I believe you; I could easily have misread the ensuing discussion in #3381. Nevertheless, so far as parsing BBC is concerned, declaring the attach tag's type as `unparsed_equals_content` when it actually takes the form of the `unparsed_content` borks the parsing of that BBC tag. The problem @PortaMx described in #3381 is solved by the combination of #3447 and #3457. You know far, far more than I do about what's going on with the editor, its JavaScript, and the attachments system in general, so I'm not going to try to tell you what ought to be done about any problems there. But f81e4bb won't work if we want BBC parsing to work correctly.

**Comment by PortaMx** [comment]
All the changes/fixes they i made on the bbc parsing (img and attach) can you see in my fork (branch portamx) .. and that works perfectly on ALL situations ...

**Comment by Sesquipedalian** [comment]
Well, by changing the tag type on the `img` and `attach` BBCode tags from `unparsed_content` to `unparsed_equals_content`, your branch "fixed" my earlier mistake by introducing another error to cancel it out. The form for an `unparsed_equals_content` tag is `[tag=...]content[/tag]`. The `img` and `attach` tags do not use that form. Instead, they use `[tag param=value]content[/tag]`. By declaring the wrong type for those two tags, your branch breaks the parameter parsing for

**Comment by Sesquipedalian** [comment]
Well, by changing the tag type on the img and attach BBCode tags from unparsed_content to unparsed_equals_content, your branch "fixed" my earlier mistake by introducing another error to cancel it out, @PortaMx.

The form for an unparsed_equals_content tag is `[tag=...]content[/tag]`. The img and attach tags do not use that form. Instead, they use `[tag param=value]content[/tag]`. So unparsed_equals_content is the wrong type for those tags.

By declaring the wrong type for those two tags, your branch breaks the parameter parsing for them. That avoided the bug in the new parameter parsing code by not parsing the parameters at all. That's why you had to add code to the validate section to parse the parameters yourself.

The real cause of the problem is the mistake I made when writing the new parameter parsing function. #3447 fixes that mistake, and the result works perfectly in every situation, too. @MissAllSunday says there were some issues with attachments appearing in the editor that f81e4bb was supposed to help solve, but after applying #3457, which reverts f81e4bb, image attachments are showing up in the editor and in previews perfectly.

**Comment by PortaMx** [comment]
Well ..I will give it a look .. but my code works perfectly on all situations .. never change a runnig code ;-)

**Comment by Sesquipedalian** [comment]
I do like that your branch also added width and height parameters to the attach tag, @PortaMx. That's a great idea. Do you mind if I take that bit of code and put in into a new pull request to merge into the main release-2.1 branch?

**Comment by PortaMx** [comment]
I tested YOUR attach code, but .. this don't works on for me ..
go back to mine version :-)

Because I see that SMF don't like my code changes (I have many done) i never more make a pull,
But you can grab the code from my repository (use the branch portamx)
I will update the branch in the next 2 hours ...

**Comment by Sesquipedalian** [comment]
> I tested YOUR attach code, but .. this don't works on for me ..
> go back to mine version :-)

Well, I don't know why it isn't working for you. Maybe you made some other changes in your branch that are interfering with it. But hey, in your branch/fork, do whatever you like. :)

**Comment by PortaMx** [comment]
I download the release branch 2 hours ago .. but the code don't work for me, because I NEED width, height, and more on the attach and img tag .. because I use Lghtbox2 as image/gallery viever ;-)

**Comment by Sesquipedalian** [comment]
Oh, I see what you meant now. You meant that it doesn't work for you in the sense of not doing what you want. I thought you meant that it didn't work for you in the sense of giving you errors. 

Width and height work for the img tag in release-2.1, but not yet for the attach tag. I'm going to borrow some of your code to add support for width and height to the attach tag.

**Comment by PortaMx** [comment]
On "borrow" .. try a look to the sceditor .. the qoute style .. on my code they looks identically as on posts ;-)
Also I fixed a lot of bugs on the smf_fileUpload.js and dropzone.js .. and mutch more ;-)

Live forum/portal on http://smf21devel.portamx.com/

**Comment by MissAllSunday** [comment]
Feline.....we've been through this over and over again...

SMF welcomes every contribution, HOWEVER there are rules and procedures every contributor needs to follow, you are either unable or unwilling to follow those procedures.

You keep on adding new code on top of old branches, this will cause you nothing but issues all the way in.

As demonstrated by the recent bug reports you have made, they all come from using old code on old outdated branches.

The PRs you made are all based on old branches, thus we are unable to merge them.  Please DO try to understand that this doesn't mean we don't want your contributions, it simply means we CANNOT accept them.

I'm really sorry but one of the essential requisites to be able to contribute to SMF is having a minimal understanding on how git works, to know how to create/update/rebase branches is a must specially when dealing with OS with a considerable amount of contributors.

**Comment by PortaMx** [comment]
Ok .. I'm not a expert on git .. but .. all two days I download the devel branch and merge this in my local copy. On this I work, commit local and Push this to my branch on GitHub. That's all ..

**Comment by Sesquipedalian** [comment]
Here's a quick tutorial on what you need to do to make usable pull request, @PortaMx. Some of this you probably already know, and some you might not.

I am going to assume that you already have a local clone of your PortaMx/SMF2.1 repository on your computer, and I will show the instructions using a Unix terminal.

In order to make clean pull requests, you'll first need to get your local copy of the release-2.1 branch synced up with the copy of that branch in the official repository. At the time I am writing this, your fork of the release-2.1 branch includes a bunch of changes that were merged in from your portamx branch, but which were never accepted into the official repository. We are going to reset those changes, and then merge in all the commits that have been made to the official version of release-2.1. Run the following commands in your terminal:

```
cd /path/to/local/copy/of/SMF2.1
git checkout release-2.1
git reset HEAD^ --hard
git push release-2.1 -f
git pull https://github.com/SimpleMachines/SMF2.1.git release-2.1
git push release-2.1
```

Now that your fork of the release-2.1 branch is up to date with the official repository, you'll be able to make clean pull requests. Here's the process to do that:
1. Make sure you are in the release-2.1 branch: `git checkout release-2.1`
2. Pull in any recent changes from the official repository: `git pull https://github.com/SimpleMachines/SMF2.1.git release-2.1`
3. **Create a new local branch for your changes**: `git branch my-new-branch`
4. Switch to your new branch: `git checkout my-new-branch`
5. Make your changes and commit them (remember to use the `-s` option on the commit)
6. Push your new local branch to your repository on github: `git push -u origin`
7. On GitHub, submit your new branch as a pull request.

By creating a separate branch for each new feature/bug fix that you want to submit, you can keep your different features cleanly separated from each other, which will make it much easier for SimpleMachines to accept and merge your pull requests. And by making sure that you pull in the latest changes from the official repository just before you create that new branch, you can make sure that you are basing your work on the latest code, which will help you to avoid merge conflicts.

I hope that helps.

**Comment by PortaMx** [comment]
Thanks, for your help, I will test that ...
